### DISCUSSION TITLE
MUI TextInput loses focus when isDirty state changes

### DISCUSSION THREAD

**Discussion by travdog95** [react-hook-form/react-hook-form#11721] [Q&A]
I need some help understanding why the firstName field loses focus when the isDirty state changes. When I remove the formState from my useForm call, the firstName field doesn't lose focus. I've tried removing the useEffect and blocker code, but that doesn't resolve the issue either. For some reason my form is unmounting and mounting when isDirty changes state. Here's the code:

import { useEffect } from "react";
import { useForm, Controller } from "react-hook-form";
import { z } from "zod";
import { zodResolver } from "@hookform/resolvers/zod";
import { unstable_useBlocker as useBlocker } from "react-router-dom";
import {
  Grid,
  Button,
  FormGroup,
  Checkbox,
  Typography,
  Box,
  FormControlLabel,
  TextField,
} from "@mui/material";

import MemberPrayers from "./MemberPrayers";
import MemberTalks from "./MemberTalks";
import MemberCallings from "./MemberCallings";
import ProfileLabel from "./ProfileLabel";
import ProfileItem from "./ProfileItem";
import ConfirmationDialog from "../../components/ConfirmationDialog";
import { calculateAge, formatDate } from "../../utils/helpers";

// Define a validation schema using zod
const memberProfileSchema = z.object({
  firstName: z.string().optional(),
  isWillingToAcceptCalling: z.boolean().optional(),
  isWillingToPray: z.boolean().optional(),
  isWillingToSpeak: z.boolean().optional(),
  contactForTithing: z.boolean().optional(),
  isServingMission: z.boolean().optional(),
});

const ProfileForm = ({ member, updateMember, meetings, callings }) => {
  const age = calculateAge(member.birthDate);
  const birthday = formatDate(member.birthDate, "LLL d");

  // Create a form using react-hook-form, using zod for validation
  const {
    control,
    handleSubmit,
    formState: { isDirty },
    reset,
  } = useForm({
    defaultValues: {
      ...member,
    },
    resolver: zodResolver(memberProfileSchema),
  });

  // Block the user from leaving the page if there are unsaved changes
  let blocker = useBlocker(
    ({ currentLocation, nextLocation }) =>
      isDirty && currentLocation.pathname !== nextLocation.pathname
  );

  // Reset the form when the member changes
  useEffect(() => {
    reset(member);
  }, [member]);

  const gender = member.gender === "M" ? "Male" : "Female";

  // Save the member data, reset the form, and unblock the user
  const handleSaveMember = (data) => {
    updateMember.mutate({ id: member._id, ...data });
    reset(data);
  };

  return (
    <Box sx={{ width: "100%" }} mt={2}>
      <form onSubmit={handleSubmit(handleSaveMember)}>
        {blocker.state === "blocked" && (
          <ConfirmationDialog
            open={blocker.state === "blocked"}
            title="Unsaved Changes!"
            message="Are you sure you want to leave without saving?"
            onConfirm={() => blocker.proceed()}
            onCancel={() => blocker.reset()}
          />
        )}

        <Grid container columnSpacing={4}>
          <Grid item xs={12} md={6}>
            <ProfileItem>
              <ProfileLabel>First Name</ProfileLabel>
              <Controller
                name="firstName"
                control={control}
                render={({ field }) => (
                  <TextField
                    {...field}
                    variant="outlined"
                    fullWidth
                    size="small"
                    disabled={updateMember.isLoading}
                  />
                )}
              />
            </ProfileItem>
            <ProfileItem>
              <ProfileLabel>Birthday (Age)</ProfileLabel>
              <Typography variant="body1">
                {birthday} ({age})
              </Typography>
            </ProfileItem>
            <ProfileItem>
              <ProfileLabel>Gender</ProfileLabel>
              <Typography variant="body1">{gender}</Typography>
            </ProfileItem>
            <ProfileItem>
              <ProfileLabel>Address</ProfileLabel>
              <Typography variant="body1">{member.address1}</Typography>
            </ProfileItem>
            <ProfileItem>
              <ProfileLabel>Activity</ProfileLabel>
              <FormGroup>
                <Controller
                  name="isWillingToPray"
                  control={control}
                  render={({ field }) => (
                    <FormControlLabel
                      control={<Checkbox {...field} name="isWillingToPray" checked={field.value} />}
                      label="Willing to Pray?"
                    />
                  )}
                />
                <Controller
                  name="isWillingToSpeak"
                  control={control}
                  render={({ field }) => (
                    <FormControlLabel
                      control={
                        <Checkbox {...field} name="isWillingToSpeak" checked={field.value} />
                      }
                      label="Willing to Speak?"
                    />
                  )}
                />
                <Controller
                  name="isWillingToAcceptCalling"
                  control={control}
                  render={({ field }) => (
                    <FormControlLabel
                      control={
                        <Checkbox
                          {...field}
                          name="isWillingToAcceptCalling"
                          checked={field.value}
                        />
                      }
                      label="Willing to Accept Calling?"
                    />
                  )}
                />
                <Controller
                  name="contactForTithing"
                  control={control}
                  render={({ field }) => (
                    <FormControlLabel
                      control={
                        <Checkbox {...field} name="contactForTithing" checked={field.value} />
                      }
                      label="Contact for Tithing?"
                    />
                  )}
                />
                <Controller
                  name="isServingMission"
                  control={control}
                  render={({ field }) => (
                    <FormControlLabel
                      control={
                        <Checkbox {...field} name="isServingMission" checked={field.value} />
                      }
                      label="Serving Mission?"
                    />
                  )}
                />
              </FormGroup>
            </ProfileItem>
          </Grid>
          <Grid item xs={12} md={6}>
            <MemberCallings member={member} callings={callings} />
            <MemberTalks member={member} meetings={meetings} />
            <MemberPrayers member={member} meetings={meetings} />
          </Grid>
          <Grid item xs={12} mt={0} pt={0}>
            <Button
              type="submit"
              variant="contained"
              color="primary"
              fullWidth
              sx={{ fontSize: "1.3rem" }}
              disabled={!isDirty}
            >
              Save Profile
            </Button>
          </Grid>
        </Grid>
      </form>
    </Box>
  );
};

export default ProfileForm;

**Comment by ddaeyongee** [comment]
how about
```
useEffect(() => {
  if (isDirty) {
    reset(member);
  }
}, [member, isDirty]);
```
